snapshots/climate/2024-01-31/climate_change_impacts.py

```diff
@@ -0,0 +1,202 @@
+#   But in the following link, they claim the data is updated every year, so it could be added to our yearly data.
+#   https://www.data.jma.go.jp/gmd/kaiyou/english/ohc/ohc_global_en.html
+# * Rutgers University Global Snow Lab also includes snow cover extent for:
+#   * Eurasia: https://climate.rutgers.edu/snowcover/files/moncov.eurasia.txt
+#   * North America (excluding Greenland): https://climate.rutgers.edu/snowcover/files/moncov.nam.txt
+# * Ice sheet mass balance from NASA EarthData. This is regularly updated, but to access it one has to manually log in.
+#   The data can be manually accessed from:
+#   https://climate.nasa.gov/vital-signs/ice-sheets/
+#   By clicking on the HTTP link. This leads to a manual log in page.
+#   Once logged in, the data is accessible via the following link:
+#   https://archive.podaac.earthdata.nasa.gov/podaac-ops-cumulus-protected/ANTARCTICA_MASS_TELLUS_MASCON_CRI_TIME_SERIES_RL06.1_V3/antarctica_mass_200204_202310.txt
+#   So, one could use this link, trying with different dates (e.g. ..._202401.txt, ..._202312.txt, ..._202311.txt),
+#   until the most recent file is downloaded.
+#   I contacted EarthData to ask if there is any way to access the latest data programmatically.
+# * Global sea level from NASA.
+#   We could get more up-to-date data on sea levels from https://sealevel.jpl.nasa.gov/
+#   but we would need to use a special library with credentials to fetch the data (and the baseline and format would
+#   probably be different).
+########################################################################################################################
+
+
+def find_date_published(snap: Snapshot) -> Optional[str]:
+    # Extract publication date for each individual origin, if possible.
+    # Otherwise, assign the current access date as publication date.
+    if snap.path.name == "sea_ice_index.xlsx":
+        # * For sea_ice_index, the date_published can be found on:
+        #   https://noaadata.apps.nsidc.org/NOAA/G02135/seaice_analysis/
+        #   Next to the file name (Sea_Ice_Index_Monthly_Data_by_Year_G02135_v3.0.xlsx).
+
+        # Extract all the text in the web page.
+        url = "/".join(snap.metadata.origin.url_download.split("/")[:-1])  # type: ignore
+        response = requests.get(url)
+        # Parse HTML content.
+        soup = BeautifulSoup(response.text, "html.parser")
+
+        # Fetch the date that is written next to the title.
+        for line in soup.text.split("\n"):
+            if "Sea_Ice_Index_Monthly_Data_by_Year" in line:
+                dates = re.findall(r"\d{2}-\w{3}-\d{4}", line)
+                if len(dates) == 1:
+                    # Format date conveniently.
+                    date = datetime.strptime(dates[0], "%d-%b-%Y").strftime("%Y-%m-%d")
+                    return date
+                else:
+                    log.warn(f"Failed to extract date_published for: {snap.path.name}")
```

Comment:
I guess you can move this line (and the other equivalent ones in the other `ifs`) to the end of the function, just before the `return` statement, so you call this once instead of four times.

Reply:
Not exactly. The function should log a warning only when handling any of the steps for which we can extract the publication date. For any other cases, it should return the default publication date (which is the access date). If this function grows more, I may rethink its logic to DRY it a bit.